javascript/net/grpc/web/grpcwebclientbase.js

```diff
@@ -123,18 +123,18 @@ class GrpcWebClientBase {
       let unaryStatus;
       let unaryMsg;
       GrpcWebClientBase.setCallback_(
-          stream, (error, response, status, metadata) => {
+          stream, (error, response, status, metadata, trigger) => {
             if (error) {
               reject(error);
-            } else if (response) {
-              unaryMsg = response;
             } else if (status) {
               unaryStatus = status;
             } else if (metadata) {
               unaryMetadata = metadata;
-            } else {
+            } else if (trigger) {
               resolve(request.getMethodDescriptor().createUnaryResponse(
                   unaryMsg, unaryMetadata, unaryStatus));
+            } else {
+              unaryMsg = response;
```

Comment:
I'm not sure we should set the `unaryMsg` on an `else {}` case — it seems a bit unsafe to set it before checking nullability (even though it might be ok based on existing callers).

Actually.. I'm curious what if you don't make this change? With the existing code, wouldn't the `stream.on('end' ...)` below eventually trigger a `callback(null, null)` call which will fall into this `else {}`clause, which will trigger the `resolve(...)` call?

Reply:
> I'm not sure we should set the `unaryMsg` on an `else {}` case — it seems a bit unsafe to set it before checking nullability (even though it might be ok based on existing callers).

This single callback has to do 5 different jobs in a function.

1. receive an error
2. receive the response value
3. receive the status
4. receive the metadata
5. be informed that we have all pieces needed to give the caller the unary response

So far, it has done this by using a giant `if ...` tree and testing for various arguments' falsiness. Supporting falsy response values would require a way to differentiate job 2 and job 5. This version declares that job 5 has `trigger = true`, while job 2 has `trigger = false`, which overall sticks to the existing theme of "one of the args is truthy, just check which one."

As for how can we ensure it's safe: `setCallback_` is private and the callback handle is kept private, so we have control over what will ever call it, and we can make sure all callers follow this new protocol in this PR.

> Actually.. I'm curious what if you don't make this change? With the existing code, wouldn't the `stream.on('end' ...)` below eventually trigger a `callback(null, null)` call which will fall into this `else {}`clause, which will trigger the `resolve(...)` call?

If we don't change the protocol:

1. A falsy response never gets assigned into `unaryMsg`. I believe our use case is sloppy-tolerant enough not to experience any problems if grpc-web gives `undefined` instead of `0` or `""` or `false` or `null`, but giving the right type is still nice.
2. The `resolve` would be called twice. I think standards compliant Promise implementations are specified to ignore extra calls, so that shouldn't break anything. But that's a relatively more obscure detail, and I think the code is easier for beginners to verify if it only ever calls `resolve` once.

Reply:
3. Actually I don't know about the order of calls to this callback. And if I did know the order, I don't know it's an order that gRPC intends to stick to. If job 2 comes before job 4, for example (I vaguely recall seeing something about trailers in gPRC? Maybe grpc-web doesn't support that though.), we'd incorrectly `resolve` before setting the unary metadata.

Reply:
Hi! Thanks for the quick reply and sorry for the slow reply.. It took me a while to wrap my head around the code here.. 😅

> As for how can we ensure it's safe: setCallback_ is private and the callback handle is kept private, so we have control over what will ever call it, and we can make sure all callers follow this new protocol in this PR.

What you said makes a lot of sense.. I guess i was kind hoping that there's an explicit signal for setting the response (for better robustness).. but given that the response is falsy here, what you have here is probably the best we can do for now :) (Especially given that the "callback" is not exactly "internal" since we reuse the same function for external callers ([here](https://github.com/grpc/grpc-web/blob/b0ea9c7c45d6f9ea5338fc61a81b47f589a91259/javascript/net/grpc/web/grpcwebclientbase.js#L103-L110)) too..)

> If we don't change the protocol:
> 1. A falsy response never gets assigned into unaryMsg. I believe our use case is sloppy-tolerant enough not to experience any problems if grpc-web gives undefined instead of 0 or "" or false or null, but giving the right type is still nice.
> 2. The resolve would be called twice. I think standards compliant Promise implementations are specified to ignore extra calls, so that shouldn't break anything. But that's a relatively more obscure detail, and I think the code is easier for beginners to verify if it only ever calls resolve once.

Thanks so much for explaining! This makes total sense.. :)

> 3, Actually I don't know about the order of calls to this callback. And if I did know the order, I don't know it's an order that gRPC intends to stick to. If job 2 comes before job 4, for example (I vaguely recall seeing something about trailers in gPRC? Maybe grpc-web doesn't support that though.), we'd incorrectly resolve before setting the unary metadata.

Yeahh i think grpc-web is not using HTTP trailers due to API limitations (see [here](https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-WEB.md)) so what you've mentioned shouldn't happen. But thanks for the careful considerations :)

Reply:
here's something we could do, let me know if this is anymore idiomatic:

we still add an extra parameter, but instead of true = "done" and false = "get the response value," we can have true = "get the response value" and false = "done"

Reply:
pushed a version where instead of a having a unaryEndOfStream flag, there's a unaryResponseReceived flag. i.e., now "there's an explicit signal for setting the response."

added doc comment for the useUnaryResponse  parameter to setCallback_